### PR TITLE
Bzero/Memset optimizations

### PR DESCRIPTION
**Pull request by AtariDreams**
bzero has been deprecated since 2008. It's best we remove it

### CODE REVIEW

**Comment by AtariDreams** [comment]
cc @compnerd

**Comment by AtariDreams** [comment]
Hopefully this change can be reflected in CoreFoundation on Apple's end too

**Comment by AtariDreams** [comment]
@swift-ci please test

**Comment by AtariDreams** [comment]
cc @rokhinip @etcwilde @compnerd

**Comment by parkera** [comment]
@swift-ci please test

**Comment by AtariDreams** [comment]
@parkera any updates?

**Comment by AtariDreams** [comment]
Any updates?

**Comment by etcwilde** [comment]
My thoughts: I personally prefer `bzero` over `memset`, just because it avoids a certain class of errors.
e.g. `memset(ptr, sizeof(type), 0);` is an unfortunately common bug that the compiler won't warn about since the types are right. Yes, you are right that `bzero` is deprecated, but it's also very unlikely that it's going anywhere. I would prefer leaving it in until it becomes an issue. Same thoughts on https://github.com/apple/swift-corelibs-libdispatch/pull/579#issuecomment-943686526.

I see that @parkera is fine with the changes going into Foundation, that should be sufficient approval. Please check out https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md to learn more about how to trigger the pull-request testing. In this case, you'll want `@swift-ci please test`.

If you are looking for starter bugs to tackle to get more familiar with the community and the Swift project, please feel free to take a look here: https://bugs.swift.org/browse/SR-15313?filter=10625. The bugs will help provide more context, making it easier to us to review your patches.

**Comment by AtariDreams** [comment]
Actually, many of the BSD distros are starting to remove support for bzero. Granted, the process for this has only just started on these distros, but it is happening.

**Comment by millenomi** [comment]
@swift-ci please test

**Comment by AtariDreams** [comment]
Done and finally ready for review/merge!

File: CoreFoundation/Collections.subproj/CFData.c [hunk 1]
```diff
@@ -283,12 +283,10 @@ static void *__CFDataAllocate(CFDataRef data, CFIndex size, Boolean clear) {
 	CFAllocatorRef allocator = __CFGetAllocator(data);
 	bytes = CFAllocatorAllocate(allocator, size, 0);
 	if (clear) memset((uint8_t *)bytes, 0, size);
+    } else if (clear) {
+    bytes = calloc(1, size);
```
Comment (stevapple) [inline CoreFoundation/Collections.subproj/CFData.c:287]:
```suggestion
        bytes = calloc(1, size);
```

File: CoreFoundation/Collections.subproj/CFData.c [hunk 2]
```diff
@@ -283,12 +283,10 @@ static void *__CFDataAllocate(CFDataRef data, CFIndex size, Boolean clear) {
...
     } else {
-        if (clear) {
-            bytes = calloc(1, size);
-        } else {
-            bytes = malloc(size);
-        }
+    bytes = malloc(size);
```
Comment (stevapple) [inline CoreFoundation/Collections.subproj/CFData.c:289]:
```suggestion
        bytes = malloc(size);
```

File: CoreFoundation/Base.subproj/CFRuntime.c [hunk 3]
```diff
@@ -481,20 +481,22 @@ CFTypeRef _CFRuntimeCreateInstance(CFAllocatorRef allocator, CFTypeID typeID, CF
...
         memory = (CFRuntimeBase *)CFAllocatorAllocate(allocator, size, 0);
-        needsClear = true;
-    }
-
-    if (NULL == memory) {
-	return NULL;
-    } else if (needsClear) {
+        if (NULL == memory) {
+            return NULL;
+        }
```
Comment (stevapple) [inline CoreFoundation/Base.subproj/CFRuntime.c:499]: Why redundant checks for `memory` here?

File: CoreFoundation/String.subproj/CFCharacterSet.c [hunk 4]
```diff
@@ -503,10 +503,10 @@ CF_INLINE uint32_t __CFCSetGetCompactBitmapSize(const uint8_t *compactBitmap) {
 
 CF_INLINE void __CFExpandCompactBitmap(const uint8_t *src, uint8_t *dst) {
     const uint8_t *srcBody = src + __kCFCompactBitmapNumPages;
-    int i;
+    unsigned int i;
     uint8_t value;
 
-    for (i = 0;i < __kCFCompactBitmapNumPages;i++) {
+    for (i = __kCFCompactBitmapNumPages;i > 0;i--) {
```
Comment (stevapple) [inline CoreFoundation/String.subproj/CFCharacterSet.c:509]:
```suggestion
    for (int i = 0; i < __kCFCompactBitmapNumPages; i++) {
```

File: CoreFoundation/String.subproj/CFCharacterSet.c [hunk 5]
```diff
@@ -503,10 +503,10 @@ CF_INLINE uint32_t __CFCSetGetCompactBitmapSize(const uint8_t *compactBitmap) {
 
 CF_INLINE void __CFExpandCompactBitmap(const uint8_t *src, uint8_t *dst) {
     const uint8_t *srcBody = src + __kCFCompactBitmapNumPages;
-    int i;
+    unsigned int i;
```
Comment (stevapple) [inline CoreFoundation/String.subproj/CFCharacterSet.c:506]:
```suggestion
```

**Review by stevapple — CHANGES_REQUESTED** [review]

File: CoreFoundation/Base.subproj/CFRuntime.c (hunk 3 above)
Comment (AtariDreams) [reply CoreFoundation/Base.subproj/CFRuntime.c:499]: Redundant? It's possible for the functions above to return NULL

Comment (stevapple) [reply CoreFoundation/Base.subproj/CFRuntime.c:499]: I was curious about the purpose of this change (which introduced duplicated codes), to eliminate the allocation of a `Boolean` variable?

Comment (AtariDreams) [reply CoreFoundation/Base.subproj/CFRuntime.c:499]: Yeah that is what happened and less error prone

**Comment by GameCubeGBA** [comment]
Any updates on this? @rokhinip

**Comment by GameCubeGBA** [comment]
> My thoughts: I personally prefer `bzero` over `memset`, just because it avoids a certain class of errors. e.g. `memset(ptr, sizeof(type), 0);` is an unfortunately common bug that the compiler won't warn about since the types are right. Yes, you are right that `bzero` is deprecated, but it's also very unlikely that it's going anywhere. I would prefer leaving it in until it becomes an issue. Same thoughts on [apple/swift-corelibs-libdispatch#579](https://github.com/apple/swift-corelibs-libdispatch/pull/579).
> 
> I see that @parkera is fine with the changes going into Foundation, that should be sufficient approval. Please check out https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md to learn more about how to trigger the pull-request testing. In this case, you'll want `@swift-ci please test`.
> 
> If you are looking for starter bugs to tackle to get more familiar with the community and the Swift project, please feel free to take a look here: https://bugs.swift.org/browse/SR-15313?filter=10625. The bugs will help provide more context, making it easier to us to review your patches.

99.9% of the time it's not going to be wrong

**Comment by rokhinip** [comment]
@parkera owns Foundation and should make the call here. Also you need CI testing.

**Comment by etcwilde** [comment]
> > My thoughts: I personally prefer `bzero` over `memset`, just because it avoids a certain class of errors. e.g. `memset(ptr, sizeof(type), 0);` is an unfortunately common bug that the compiler won't warn about since the types are right. Yes, you are right that `bzero` is deprecated, but it's also very unlikely that it's going anywhere. I would prefer leaving it in until it becomes an issue. Same thoughts on [apple/swift-corelibs-libdispatch#579](https://github.com/apple/swift-corelibs-libdispatch/pull/579).

> 99.9% of the time it's not going to be wrong

I'm not sure I follow. `memset(ptr, sizeof(type), 0)` is probably not what someone wanted. They probably wanted `memset(ptr, 0, sizeof(type))`. It won't crash per-se, but it won't zero out the memory either.

**Comment by AtariDreams** [comment]
@etcwilde @parkera ready

**Comment by AtariDreams** [comment]
@drexin Thoughts on this?

**Pull request closed without merging**